esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,349 @@
+                                            UnsignedShortArray & returnAttribList,
+                                            StringArray & ids,
+                                            IArrayOf<CAttributeFilter> & arrayAttributeFilter,
+                                            unsigned __int64 minVersion,
+                                            unsigned __int64 & maxTimestamp)
+{
+    unsigned maxDepth = nfilter.getDepth();
+    Owned<CAttributeFilter> tmpForAttribSearch = new CAttributeFilter;
+    bool limitToIdList = ids.ordinality()>0;
+    bool returnSpecifyReturnAttribs = returnAttribList.ordinality()>0;
+    bool filterNestedScopeTypes = nestedScopeTypes.ordinality()>0;
+    bool filterByAttributeValue = arrayAttributeFilter.ordinality();
+
+    DBGLOG("Id list count: %d", ids.ordinality());
+    DBGLOG("ReturnAttribute  list count: %d", returnAttribList.ordinality());
+    DBGLOG("filterNestedScopeTypes count: %d", nestedScopeTypes.ordinality());
+    DBGLOG("AttributeFilter count: %d", arrayAttributeFilter.ordinality());
+
+    ids.sortAscii();      // Will need to lookup up 'id' in ids (for bSearch)
+    IArrayOf<IEspWUResponseAttribute> attribs;
+    SCMStringBuffer currentScope,currentScopeType,currentId;
+    unsigned statsCount = 0, statsAdded = 0;
+    ForEach(statsIter)
+    {
+        ++statsCount;
+        IConstWUStatistic & wuStat = statsIter.query();
+
+        SCMStringBuffer wuScope,  wuId;
+        wuStat.getScope(wuScope);
+        getScopeIdFromScopeName(wuId, wuScope);
+
+        if (limitToIdList && ids.bSearch(wuId.str(),funcCompareString)==NotFound)
+              continue;
+
+        if (maxDepth && getScopeDepth(wuScope) > maxDepth)
+            continue;
+
+        if (filterNestedScopeTypes && nestedScopeTypes.bSearch(wuStat.getScopeType(),funcCompareUnsignedShort)==NotFound)
+              continue;
+
+        if (filterByAttributeValue)
+        {
+            tmpForAttribSearch->setStatisticKind(wuStat.getKind());
+            aindex_t pos = arrayAttributeFilter.bSearch(tmpForAttribSearch,funcCompareAttribFilter);
+            if (pos!=NotFound)
+            {
+                CAttributeFilter & attributeFilter = arrayAttributeFilter.item(pos);
+                unsigned __int64 graphRawValue = wuStat.getValue();
+
+                if (attributeFilter.hasExactValue())
+                {
+                    if (graphRawValue != attributeFilter.getExactValue()) continue;
+                }
+                else
+                {
+                    if (attributeFilter.hasMinValue() && graphRawValue < attributeFilter.getMinValue()) continue;
+                    if (attributeFilter.hasMaxValue() && graphRawValue > attributeFilter.getMaxValue()) continue;
+                }
+            }
+        }
+        // Check if scope name has changed. If so, need to return the list attributes for previous scope
+        if (strcmp(currentScope.str(),wuScope.str())!=0)
```

Comment:
I originally though currentScope should possibly be called previousScope.
It would make more sense for this code to be right at the head of the loop, since it is effectively flushing out any previous values before processing the next item.
If the code was at the head the current name is more logical, since it is then the currentScope for all the subsequent processing.